File: services/api/app/domain/direct.py
"""Nhắn riêng: a two-person context (ADR-0021 §2.5).

A direct message is not a second kind of conversation with tables of its own.
It is a `context` whose `kind` is `pair`, so messages, photos, reactions, read
marks, the companion and the theme all work on it unchanged. What differs is
who may open one and which doors are closed once it exists, and both of those
rules live here so the service, the SQL and the tests read one spelling.

Pure: strings and booleans in, strings and booleans out. The friendship module
already knows how to order two ids; this module reuses that ordering so the
unique key in the database and `uq_friend_edge_live` agree on which of two
people comes first.
"""

from __future__ import annotations

from collections.abc import Iterable

from app.domain.friendship import pair_key as _ordered_pair

KIND_GROUP = "group"
KIND_PAIR = "pair"
#: Every value `contexts.kind` may hold; the CHECK constraint is the second
#: spelling of this tuple.
KINDS = (KIND_GROUP, KIND_PAIR)

#: What a pair is called when nobody's name is available to call it by. A
#: word, never an id: two unnamed people must not read as one hexadecimal.
ANONYMOUS_COUNTERPART = "Thành viên"

#: Doors that only make sense against a roster somebody curates. A pair has
#: exactly two members for as long as it exists, so each of these answers
#: `not_a_group` there (§2.5.5). The list is closed on purpose and read by the
#: tests that walk every door: money, outings, votes and memories are absent
#: because splitting a bill between two friends is a real thing to do.
ROSTER_ONLY_DOORS = (
    "invite_context_member",
    "accept_context_membership",
    "leave_context",
    "set_context_member_role",
    "create_outing_invite",
    "rename_context",
)


def pair_key(a: str, b: str) -> str:
    """The value `uq_contexts_pair_key` guards: both ids, ordered, joined by `:`.

    Ordered by `friendship.pair_key`, so A opening a conversation with B and B
    opening one with A compute the same key and land on the same row. Raises
    `FriendshipError("SELF_EDGE")` for one person twice and `PERSON_REQUIRED`
    for an empty id -- the same refusals a friend request gives.
    """
    first, second = _ordered_pair(a, b)
    return f"{first}:{second}"


def is_pair(kind: str | None) -> bool:
    return kind == KIND_PAIR


def is_kind(value: object) -> bool:
    return isinstance(value, str) and value in KINDS


def can_open(
    *, is_friend: bool, other_exists: bool = True, other_deleted: bool = False
) -> bool:
    """Whether one person may open (or re-open) a pair with another.

    Friendship is the consent step (§2.5.2): there is no «accept conversation»
    because accepting the friend request already was that. A person who does
    not exist, or whose account is gone (ADR-0023, L5), cannot be written to.
    The caller collapses every `False` into one 404 -- this function says
    *whether*, never *why*, so it cannot become an oracle by accident.
    """
    return bool(is_friend) and bool(other_exists) and not bool(other_deleted)


def counterpart_of(member_ids: Iterable[str], me: str) -> str | None:
    """The one member of a pair who is not `me`; None when the roster is not a
    pair's roster (zero or several other people)."""
    others = [member for member in member_ids if member != me]
    if len(others) != 1:
        return None
    return others[0]


def display_name_for(
    kind: str | None, stored_name: str, counterpart_name: str | None
) -> str:
    """What the reader sees at the top of the conversation.

    A group is called what its members called it. A pair has no name of its
    own -- `display_name` is stored empty (§2.5.6) -- so the reader sees the
    other person's name, derived on every read and never written back. When
    that name is missing the fallback is a word, not an id.
    """
    if not is_pair(kind):
        return stored_name
    if counterpart_name:
        return counterpart_name
    return ANONYMOUS_COUNTERPART

File: services/core/internal/domain/direct/direct.go
// Package direct ports app/domain/direct.py (ADR-0021 §2.5): whether a context
// is a pair, who the other person of a pair is, what the reader sees as its
// name, and who may open one.
//
// POST /contexts, PATCH /contexts/{id} and GET /contexts/{id} answer through
// `_context_response`, which names a pair after the other member; the roster
// doors refuse a pair through `_require_group_kind` (W3,
// scripts/render_domain_w3_goldens.py). POST /people/{person_id}/dm opens a
// pair with PairKey and CanOpen (W10, scripts/render_domain_w10_goldens.py).
package direct

import (
	"slices"

	"mobile/services/core/internal/domain/friendship"
)

// Kinds of context, KIND_GROUP and KIND_PAIR.
const (
	KindGroup = "group"
	KindPair  = "pair"
)

var kinds = [...]string{KindGroup, KindPair}

// Kinds returns KINDS: every value contexts.kind may hold.
func Kinds() []string { return slices.Clone(kinds[:]) }

var rosterOnlyDoors = [...]string{
	"invite_context_member",
	"accept_context_membership",
	"leave_context",
	"set_context_member_role",
	"create_outing_invite",
	"rename_context",
}

// RosterOnlyDoors returns ROSTER_ONLY_DOORS: the service doors that answer
// not_a_group on a pair.
func RosterOnlyDoors() []string { return slices.Clone(rosterOnlyDoors[:]) }

// AnonymousCounterpart is ANONYMOUS_COUNTERPART: what a pair is called when
// nobody's name is available. A word, never an id.
const AnonymousCounterpart = "Thành viên"

// IsPair is is_pair for a str kind. The service always passes the stored
// kind, which is never None.
func IsPair(kind string) bool { return kind == KindPair }

// CounterpartOf is counterpart_of: the one member id that is not me, found
// only when exactly one entry of memberIDs differs from me. Entries equal to
// me are skipped however many there are; two entries naming the same other
// person count as two.
func CounterpartOf(memberIDs []string, me string) (string, bool) {
	other, found := "", 0
	for _, member := range memberIDs {
		if member == me {
			continue
		}
		other = member
		found++
	}
	if found != 1 {
		return "", false
	}
	return other, true
}

// DisplayNameFor is display_name_for. counterpartName nil is None; an empty
// name is falsy in Python and falls back to AnonymousCounterpart as well.
func DisplayNameFor(kind, storedName string, counterpartName *string) string {
	if !IsPair(kind) {
		return storedName
	}
	if counterpartName != nil && *counterpartName != "" {
		return *counterpartName
	}
	return AnonymousCounterpart
}

// IsKind is is_kind: a str that is one of Kinds. Any other type answers false,
// as isinstance does.
func IsKind(value any) bool {
	kind, ok := value.(string)
	return ok && slices.Contains(kinds[:], kind)
}

// PairKey is pair_key: both ids ordered by friendship.PairKey and joined by
// ":", the value uq_contexts_pair_key guards. The refusals are friendship's:
// *friendship.FriendshipError with PERSON_REQUIRED for an empty id and
// SELF_EDGE for one person twice.
func PairKey(a, b string) (string, error) {
	pair, err := friendship.PairKey(a, b)
	if err != nil {
		return "", err
	}
	return pair[0] + ":" + pair[1], nil
}

// CanOpen is can_open: a friend, who exists, whose account has not ended. The
// Python defaults are otherExists true and otherDeleted false.
func CanOpen(isFriend, otherExists, otherDeleted bool) bool {
	return isFriend && otherExists && !otherDeleted
}
